### PR TITLE
Ensure queues are not started when they are supposed to be paused or stopped

### LINKED ISSUE #2454: SetUpFixture not respecting NonParallelizable tag on TestFixtures.
It seems since version 3.8.0, the inclusion of a SetUpFixture causes issues when trying to run multiple fixtures marked as parallelizable but some fixtures marked as non-parallelizable.

I have parallelism set at the assembly level with the following in my AssemblyInfo.cs:
`[assembly: Parallelizable(ParallelScope.Fixtures)]`

Here's my test setup:
```
namespace Tests
{
	[SetUpFixture]
	public class TestsSetUpFixture
	{

	}

	[TestFixture]
	public class TestFixture1
	{
		[Test]
		public void TestFixture1_Test()
		{
			Thread.Sleep(1000);
		}
	}

	[TestFixture]
	[NonParallelizable]
	public class TestFixture2
	{
		[Test]
		public void TestFixture2_Test()
		{
			Thread.Sleep(1000);
		}
	}

	[TestFixture]
	public class TestFixture3
	{
		[Test]
		public void TestFixture3_Test()
		{
			Thread.Sleep(1000);
		}
	}
}
```

All three test fixtures get run in parallel, I would expect 1 and 3 to run in parallel whilst 2 is run separately. This is the behaviour when running under version 3.7.1, however upgrading to 3.8 causes them all to run in parallel (the same is also true under 3.8.1).

We get the same problem if you remove the parallelism setting from AssemblyInfo and instead mark TestFixture1 and TestFixture3 as [Parallelizable(ParallelScope.Self)].

As work around, for now,  you can change the ParallelScope in AssemblyInfo to Children instead of Fixtures and the behaviour returns to what is expected where all TestFixture 2 waits for the other 2 fixtures to finish executing. If you have more than 1 test in a fixture  you do end up with tests within fixtures running in parallel; this can be further fixed by adding the ParallelScope.Self above the fixtures which need to run in parallel.

### PR DESCRIPTION
**Pull request by CharliePoole**
Fixes #2454 and perhaps some others. We should check after merging.

### CODE REVIEW

File: src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs [hunk 1]
```diff
@@ -242,6 +242,11 @@ private void RestoreQueues()
 
         private void OnEndOfShift(object sender, EventArgs ea)
         {
+            // This shift ended, so see if there is work in any other
+            // TODO: Temp fix - revisit later when object model is redesigned.
+            if (StartNextShift())
+                return;
+
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs]: This change is an "extra" not related to the issue I'm fixing. But it's an obvious error. Without it, we would restore all the queues, possibly losing work items not yet started, when the shift starts. For example, if a NP fixture has parallel tests, they could be waiting in the parallel queue to be run when the non-p shift ends!

More work is needed here - I just fixed the obvious.

File: src/NUnitFramework/framework/Internal/Execution/WorkItemQueue.cs [hunk 2]
```diff
@@ -322,13 +322,16 @@ public void Pause()
...
+            if (isRunning)
+                Pause();
 
             _savedState.Push(new SavedState(this));
 
             InitializeQueues();
 
-            Start();
+            if (isRunning)
+                Start();
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Execution/WorkItemQueue.cs]: This fixes the cause of the problem. When saving the state of queues, we started all the new queues running, not just those that were already running under the control of a `WorkShift`.

**Comment by rprouse** [comment]
Looks good so far. I am just finishing for the night, so will pull in the morning and run through some tests before final approval 👍

**Comment by ChrisMaddock** [comment]
Could I suggest adding the guard added in #2445 to `RestoreQueues()` - to get additional protection while we're in this intermediate state before refactoring? I think the fix in general is good - but if we've missed any other edge cases, this may make diagnosis easier.

https://github.com/nunit/nunit/pull/2445/files#diff-5c2e53665a6b5117adeb8d5d162e4726R227

**Comment by rprouse** [comment]
I created a test project based on the tests that I was using for #2445 and they all run to completion now. I am going to add in the tests from #2454 and check those, but I think this is good to go although I agree the guard may be helpful. I will leave that to you to add and merge when ready. Once this is merged, I will continue to add to my test project to see what other related issues are fixed.

If anyone wants to play with it, the project is at https://github.com/rprouse/NUnit.Parallelism.Integration.Tests

**Comment by rprouse** [comment]
I have verified this fixes #2454 and #2438 👍 

I think it will also fix #2395

**Comment by rprouse** [comment]
This **does not** fix #2388.

Testing #2426, I now get a `NullReferenceException` on 

```
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItemQueue.IsEmpty.get()
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkShift.HasWork.get()
nunit.framework.dll!NUnit.Framework.Internal.Execution.ParallelWorkItemDispatcher.StartNextShift()
nunit.framework.dll!NUnit.Framework.Internal.Execution.ParallelWorkItemDispatcher.OnEndOfShift(object sender, System.EventArgs ea)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkShift.StartWorkers.AnonymousMethod__26_1(NUnit.Framework.Internal.Execution.TestWorker s, NUnit.Framework.Internal.Execution.WorkItem ea)
nunit.framework.dll!NUnit.Framework.Internal.Execution.TestWorker.TestWorkerThreadProc()
```
